raft: remove MsgApp flow control management dependence on heartbeats

iskettaneh opened the pull request:
This commit does the following:

Instead of unpausing the follower and maybe sending a MsgApp when receiving a heartbeat response, it does the same thing but on heartbeat timeout if the follower is supporting the leader.

NOTE: future PRs will stop heartbeats all together if storeliveness is enabled.

Fixes: #130493

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/130532)
<!-- Reviewable:end -->

iskettaneh committed:
raft: remove MsgApp flow control management dependence on heartbeats

This commit does the following:

Instead of unpausing the follower and maybe sending a MsgApp when
receiving a heartbeat response, it does the same thing but on heartbeat
timeout if the follower is supporting the leader.

Fixes: #130493

Release note: None

pav-kv commented on pkg/raft/raft_paper_test.go:
```diff
@@ -132,6 +132,8 @@ func TestLeaderBcastBeat(t *testing.T) {
 			sort.Sort(messageSlice(msgs))
 			if storeLivenessEnabled {
 				assert.Equal(t, []pb.Message{
+					{From: 1, To: 2, Term: 1, Type: pb.MsgApp, Entries: r.raftLog.allEntries()},
```
Could you update the test description?

Also, consider putting a `require.Empty(t, r.readMessages())` before the ticks. Maybe even after each tick until the last one. And so `hi > 1` would be good :)

pav-kv commented on pkg/raft/raft_test.go:
```diff
@@ -1204,6 +1211,59 @@ func TestHandleHeartbeatResp(t *testing.T) {
+
+	// On heartbeat timeout, the leader sends a MsgApp.
+	for ticks := sm.heartbeatTimeout; ticks > 0; ticks-- {
+		sm.tickHeartbeat()
```
`sm.tick()`?

pav-kv commented on pkg/raft/raft_test.go:
```diff
@@ -1977,42 +2037,76 @@ func TestLeaderAppResp(t *testing.T) {
+					{From: 1, To: 2, Term: 2, Type: pb.MsgFortifyLeader},
+					{From: 1, To: 3, Term: 2, Type: pb.MsgFortifyLeader},
+					{From: 1, To: 3, Term: 2, LogTerm: 2, Index: 1011, Commit: 1000, Match: 1011,
+						Type: pb.MsgApp},
```
nit: place `Type` before `LogTerm`?

pav-kv commented on pkg/raft/tracker/progress.go:
```diff
@@ -106,12 +106,13 @@ type Progress struct {
+	// MsgAppProbesPaused is used when the MsgApp flow to a node is throttled.
+	// This happens in StateProbe, or StateReplicate with saturated Inflights. In
+	// both cases, we need to continue sending MsgApp once in a while to guarantee
+	// progress, but we only do so when MsgAppProbesPaused is false (it is reset
+	// on the next heartbeat timeout if the follower's store supports the leader's
+	// store), to not overflow the receiver. See IsPaused() and
```
While here, could you break this sentence in two? This nesting complicates reading a bit.

```golang
// ... but we only do so when MsgAppProbesPaused is false, to avoid spinning.
// The flag is reset on the next heartbeat timeout ...
```

Also, while we support both modes, should we retain/reword the previous bit about "heartbeat response"?

pav-kv commented on pkg/raft/testdata/snapshot_succeed_via_app_resp.txt:
```diff
@@ -60,7 +60,7 @@ INFO newRaft 3 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastter
 tick-heartbeat 1
 ----
-ok
+DEBUG ignore sending snapshot to 3 since it is not recently active
```
Looks weird. We've constructed a snapshot (probably because of the `compact` we weren't able to construct a `MsgApp`) and were trying to send it, but dropped. It would be good to not construct it in the first place then :) Consider leaving a TODO, don't need to get to the bottom of it now.

pav-kv commented on pkg/raft/testdata/slow_follower_after_compaction.txt:
```diff
@@ -96,24 +96,49 @@ compact 1 17
-# Trigger a round of empty MsgApp "probe" from leader. It will reach node 3
-# which will reply with a rejection MsgApp because it sees a gap in the log.
-# Node 1 will reset the MsgApp flow and send a snapshot to catch node 3 up.
+# Trigger a heartbeat timeout to allow the leader to detect that it can't send a
+# MsgApp to node 3 because it wil get rejected. Instead, the leader should send
+# a snapshot.
```
"because it will get rejected" -> "because the relevant part of the log is already compacted".

pav-kv commented on pkg/raft/testdata/lagging_commit.txt:
```diff
@@ -126,48 +126,48 @@ tick-heartbeat 1
-# leader sent the up-to-date commit index in the heartbeat message.
-#
-# See https://github.com/etcd-io/raft/issues/138 which aims to fix this.
+1->3 MsgApp Term:1 Log:1/13 Commit:13
```
Does this PR change the default behaviour somehow, or why are so many data-driven tests affected? Would it be possible to have >= 2 commits:

1. The defaults are intact. The introduction of the new code should have close to zero impact on the tests, and they should still pass.
2. Migrate the tests to also test the new behaviour (e.g. enable the new behaviour in one/few tests at a time).

pav-kv commented on pkg/raft/testdata/heartbeat_resp_recovers_from_probing.txt:
```diff
@@ -47,28 +47,32 @@ tick-heartbeat 1
   2->1 MsgHeartbeatResp Term:1 Log:0/0
+  2->1 MsgAppResp Term:1 Log:0/11 Commit:11
   3->1 MsgHeartbeatResp Term:1 Log:0/0
 > 1 handling Ready
```
Why are we sending another `MsgApp` below? There hasn't been a tick, and everything seems in flight. Is the the `MsgHeartbeatResp` logic mixing with the liveness-based one?

iskettaneh committed:
raft: replace heartbeat steps with ticks in unit tests

This commit lays the foundation for heartbeat changes with store
liveness. It replaces the direct heartbeat steps with ticks that cause
a heartbeat timeout. Also, it branches out the main heartbeat tests to
support both store liveness enabled and disabled.

This will be helpful when we stop heartbeats and rely on heartbeat
timeouts to carry out the MsgApp flow control.

Epic: None

Release note: None

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130532)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @nvanbenschoten, and @pav-kv)

___
*[`pkg/raft/tracker/progress.go` line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130532#gh-1781642339:-O84QDG60YmXyFj6gXw2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee9e30ede69a4217369a97c8e460f3df68c3d6b/pkg/raft/tracker/progress.go#L114)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

While here, could you break this sentence in two? This nesting complicates reading a bit.

```golang
// ... but we only do so when MsgAppProbesPaused is false, to avoid spinning.
// The flag is reset on the next heartbeat timeout ...
```

Also, while we support both modes, should we retain/reword the previous bit about "heartbeat response"?
</blockquote></details>

Done.
___
*[`pkg/raft/testdata/heartbeat_resp_recovers_from_probing.txt` line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130532#gh-1781719641:-O84R-g98FiX4gp_2ETv:bvz34e) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee9e30ede69a4217369a97c8e460f3df68c3d6b/pkg/raft/testdata/heartbeat_resp_recovers_from_probing.txt#L77)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Why are we sending another `MsgApp` below? There hasn't been a tick, and everything seems in flight. Is the `MsgHeartbeatResp` logic mixing with the liveness-based one?
</blockquote></details>

Because right now both code paths are active. Send a MsgApp on heartbeat timeout, and on MsgHeartbeatResp. This will change very soon when we stop MsgHeartbeat all together if store liveness is enabled.
___
*[`pkg/raft/testdata/lagging_commit.txt` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130532#gh-1781657664:-O84RCHw-CZ5dPGTGTM_:bn0jto1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee9e30ede69a4217369a97c8e460f3df68c3d6b/pkg/raft/testdata/lagging_commit.txt#L137)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Does this PR change the default behaviour somehow, or why are so many data-driven tests affected? Would it be possible to have >= 2 commits:

1. The defaults are intact. The introduction of the new code should have close to zero impact on the tests, and they should still pass.
2. Migrate the tests to also test the new behaviour (e.g. enable the new behaviour in one/few tests at a time).
</blockquote></details>

Discussed offline, I will split this into 2 commits. The first one sets up the unit tests w/ and wo/ store liveness, and the second one does the logic change and change the tests again based on the logic change.
___
*[`pkg/raft/testdata/slow_follower_after_compaction.txt` line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130532#gh-1781652005:-O84Rcr38slud0gwPfiJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee9e30ede69a4217369a97c8e460f3df68c3d6b/pkg/raft/testdata/slow_follower_after_compaction.txt#L101)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

"because it will get rejected" -> "because the relevant part of the log is already compacted".
</blockquote></details>

Done.
___
*[`pkg/raft/testdata/snapshot_succeed_via_app_resp.txt` line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130532#gh-1781645962:-O84SOACEjqLb_Pnf8Iq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee9e30ede69a4217369a97c8e460f3df68c3d6b/pkg/raft/testdata/snapshot_succeed_via_app_resp.txt#L63)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Looks weird. We've constructed a snapshot (probably because of the `compact` we weren't able to construct a `MsgApp`) and were trying to send it, but dropped. It would be good to not construct it in the first place then :) Consider leaving a TODO, don't need to get to the bottom of it now.
</blockquote></details>

Done.
___
*[`pkg/raft/node_test.go` line 770 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130532#gh-1781601085:-O84TkXo4_666COcgSiO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee9e30ede69a4217369a97c8e460f3df68c3d6b/pkg/raft/node_test.go#L770)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Ouch, this (pre-existing) code is pretty messed up.

Could we `raft.tick()` instead of `raft.tickHeartbeat()` in it, so that it's slightly more high-level? Ideally we would want to call `RawNode.Tick()`, but it looks like this test doesn't know about it.
</blockquote></details>

Done.
___
*[`pkg/raft/raft_paper_test.go` line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130532#gh-1781617741:-O84V-MN0OMTnBpfDCnC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee9e30ede69a4217369a97c8e460f3df68c3d6b/pkg/raft/raft_paper_test.go#L135)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Could you update the test description?

Also, consider putting a `require.Empty(t, r.readMessages())` before the ticks. Maybe even after each tick until the last one. And so `hi > 1` would be good :) 
</blockquote></details>

Done.
___
*[`pkg/raft/raft_test.go` line 1229 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130532#gh-1781629303:-O84VDKiDwVmfseE_Nrk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee9e30ede69a4217369a97c8e460f3df68c3d6b/pkg/raft/raft_test.go#L1229)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

`sm.tick()`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> Reviewed 11 of 13 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130532)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 895 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8H4H7b8Mc9rSWFgYjk:-O8H4H7b8Mc9rSWFgYjl:b-k061eq) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/raft.go#L895)):*
> ```Go
> 	r.trk.Visit(func(id pb.PeerID, pr *tracker.Progress) {
> 		if r.id == id {
> 			// NB: the leader doesn't send MsgAppResp to itself here. This means that
> ```

We're sending `MsgApp`s here though, not `MsgAppResp`s, so maybe the wording could be improved.

Since we're only sending `MsgApp` here, this code could only update `SentCommit`. The `MatchCommit` is not updated for a different reason - there is nothing that triggers a self-directed `MsgAppResp` except `MsgStorageAppend`s.
___
*[`pkg/raft/raft.go` line 898 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8H3-0hD4ceVCf4LWUr:-O8H3-0hD4ceVCf4LWUs:btxusbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/raft.go#L898)):*
> ```Go
> 			// the leader will not have a chance to update its own
> 			// MatchCommit/SentCommit. That is fine because the leader doesn't use
> 			// MatchCommit/SentCommit for itself. It only uses the followers' values.
> ```

nit before I forget: we often say "followers" while actually meaning a broader class. For example, the leader also sends appends to learners and tracks progress equally.

So I perhaps would drop "It only uses the followers' values", which is self-evident. Or we also often use "peers" word in this package instead, when talking broadly.
___
*[`pkg/raft/raft.go` line 905 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8HSHo_3Qt5k2y9dICS:-O8HSHo_3Qt5k2y9dICT:b-gtvczb) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/raft.go#L905)):*
> ```Go
> 			// If the follower's store isn't providing active store liveness support
> 			// to the leader's store, or it is but the leader isn't hearing about it,
> 			// we don't need to send a MsgApp.
> ```

nit: It's not like we don't need to - we do need to send it to catch up the peer :) But we don't send it because the peer does not appear live.

`we don't need to send a MsgApp` -> `don't send a MsgApp`
___
*[`pkg/raft/tracker/progress.go` line 119 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8H5_SF5tveoWAaTcvC:-O8H5_SF5tveoWAaTcvD:bxzsahs) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/tracker/progress.go#L119)):*
> ```Go
> 	// MsgAppProbesPaused is reset on the next MsgHeartbeatResp from the follower,
> 	// or on next heartbeat timeout if the follower's store supports the leader's
> 	// store)
> ```

nit: paren -> period
___
*[`pkg/raft/testdata/async_storage_writes_append_aba_race.txt` line 392 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8H61Iu49W1qEPRF1I1:-O8H61Iu49W1qEPRF1I2:bx2xd9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/testdata/async_storage_writes_append_aba_race.txt#L392)):*
> ```help
> 
> # Step 7: before the new entries reach node 1, it hears of the term change
> # through a MsgFortifyLeader/MsgApp and persists the new term. Node 1 then
> ```

It's still "heartbeat" , looking at messages below. But I understand that heartbeats will eventually be removed.

Maybe it would be prudent to not specify in this comment which message type exactly does it, just broadly say "hears of the term change and persists the new term".
___
*[`pkg/raft/testdata/heartbeat_resp_recovers_from_probing.txt` line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130532#gh-1781719641:-O8H7kvf7VwJez90vQle:bcq538s) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee9e30ede69a4217369a97c8e460f3df68c3d6b/pkg/raft/testdata/heartbeat_resp_recovers_from_probing.txt#L77)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Because right now both code paths are active. Send a MsgApp on heartbeat timeout, and on MsgHeartbeatResp. This will change very soon when we stop MsgHeartbeat all together if store liveness is enabled.
</blockquote></details>

Is there a benefit in mixing the two behaviours?

I'm still wondering why we can't make it an either/or behaviour. For example in `case MsgHeartbeatResp:` handler, could we:

```golang
if !r.fortificationTracker.FortificationEnabled() {
	pr.MsgAppProbesPaused = false
	r.maybeSendAppend(m.From)
}
```

Could place this as a helper next to `maybeUnpauseAndBcastAppend`, to make it clear that these two are mutually exclusive behaviours.

Same for the `RecentlyActive` field updates.
___
*[`pkg/raft/testdata/lagging_commit_no_store_liveness_support.txt` line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8HOveJ7aDxkFHWGV3d:-O8HOveJ7aDxkFHWGV3e:b-83sjo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/testdata/lagging_commit_no_store_liveness_support.txt#L132)):*
> ```help
> 
> # Although the leader knows that the follower 3 is lagging behind it doesn't
> # send a MsgApp because follower 3 doesn't support the leader.
> ```

Nice.
___
*[`pkg/raft/testdata/lagging_commit_no_store_liveness_support.txt` line 137 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8HPNP21ZQCrh1eZfTe:-O8HPNP21ZQCrh1eZfTf:bpbpz95) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/testdata/lagging_commit_no_store_liveness_support.txt#L137)):*
> ```help
> Ready MustSync=false:
> Messages:
> 1->2 MsgHeartbeat Term:1 Log:0/0
> ```

I wouldn't expect heartbeats here though. Is this the next change - disabling `MsgHeartbeat` if store liveness is used?
___
*[`pkg/raft/raft_flow_control_test.go` line 156 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8HWuNv2CcMlQQlw-Hw:-O8HWuNv2CcMlQQlw-Hx:b-6y2c7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/raft_flow_control_test.go#L156)):*
> ```Go
> 							ms[2].Type != pb.MsgApp || len(ms[2].Entries) != 0 {
> 							t.Fatalf("#%d.%d: len(ms) == %d, want 3 messages including one empty MsgApp",
> 								tt, i, len(ms))
> ```

This many-fold nesting doesn't look good. Maybe this test could be cleaned up before/after being modified. Not necessary to do in this PR, just a thought. It's generally a good practice - always leave the place cleaner than it was before.
___
*[`pkg/raft/raft_test.go` line 2113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8GyXMB2i6ufxqIkFF7:-O8GyXMCEs1MeoGgC9e2:bsa7k84) ([raw file](https://github.com/cockroachdb/cockroach/blob/6121b09099008d937b2408c9f4837d0dd1797beb/pkg/raft/raft_test.go#L2113)):*
> ```Go
> 		// do a heartbeat
> 		for j := 0; j < r.heartbeatTimeout; j++ {
> 			r.Step(pb.Message{From: 1, To: 1, Type: pb.MsgBeat})
> ```

I wonder why this code needed to send a heartbeat `heartbeatTimeout` times. That doesn't seem right, perhaps it only worked because the `heartbeatTimeout == 1`, set in `newTestRaft` call.

The new code makes more sense.
___
*[`pkg/raft/raft_test.go` line 2122 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8GxaFA9t7PO2xxfLDb:-O8GxaFA9t7PO2xxfLDc:b-41ps4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4be3c73e83a1ea66442e9c0585cdfa40a3fea90/pkg/raft/raft_test.go#L2122)):*
> ```Go
> 	testutils.RunTrueAndFalse(t, "store-liveness-enabled",
> 		func(t *testing.T, storeLivenessEnabled bool) {
> 			testOptions := emptyTestConfigModifierOpt()
> ```

Pre-existing, but not a big fan of this verbose construct. Can it be done as:

```golang
r := newTestRaft(..., newTestMemoryStorage(...),
	withFortification(storeLivenessEnabled))
```
___
*[`pkg/raft/raft_test.go` line 4084 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8H-LWK3bLpWJrLhXij:-O8H-LWK3bLpWJrLhXik:blyot67) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4be3c73e83a1ea66442e9c0585cdfa40a3fea90/pkg/raft/raft_test.go#L4084)):*
> ```Go
> func (nw *network) tick(p *raft) {
> 	p.tick()
> 	p.advanceMessagesAfterAppend()
> ```

`advanceMessagesAfterAppend` sounds like a thing that needs to be done after handling `Ready` and storing the log. Calling on tick is not necessarily right. But this harness is really confusing, would be great to clean it up at some point.

Looks like `p.readMessages()` calls into `p.advanceMessagesAfterAppend()`. Do we need to call both, or can just call the former?
___
*[`pkg/raft/raft_test.go` line 1225 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8HTvks1UCCY6Lmljql:-O8HTvks1UCCY6Lmljqm:b-yoiveg) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/raft_test.go#L1225)):*
> ```Go
> 	sm.fortificationTracker.RecordFortification(pb.PeerID(2), 1)
> 	sm.fortificationTracker.RecordFortification(pb.PeerID(3), 1)
> 	sm.raftLog.commitTo(LogMark{Term: 3, Index: sm.raftLog.lastIndex()})
> ```

I think the leader will be already at `Term: 4` at this time, because `becomeCandidate()` bumps the term, and then `becomeLeader()` appends a new entry at the new term, which bumps the `LogMark`'s term.

Unfortunately, we don't have good [safety checks](https://github.com/cockroachdb/cockroach/blob/290f68fb271d99f86faf7a7e7b1cb4087c7e8c65/pkg/raft/log.go#L364-L368) yet that would catch this here, but let's make sure we populate things correctly to save debugging later.

For now, can replace this with:

```golang
sm.raftLog.commitTo(sm.raftLog.unstable.mark())
```
___
*[`pkg/raft/raft_test.go` line 1238 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8HVl1s1YPBt22LbTFh:-O8HVl1s1YPBt22LbTFi:btry1mn) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/raft_test.go#L1238)):*
> ```Go
> 
> 	// On another heartbeat timeout, the leader sends a MsgApp.
> 	for ticks := sm.heartbeatTimeout; ticks > 0; ticks-- {
> ```

Seeing that this repeated tick pattern is used a lot, consider adding a helper (maybe make it test-only, by placing in a `_test.go` file) like:

```golang
func (r *raft) tickTimes(times int) {
	// this loop
}
```

Or could leave a TODO for later.


<!-- Sent from Reviewable.io -->

iskettaneh committed:
raft: clean up nested message checks in raft_flow_control_test

Nested message checks in raft_flow_control_test make it hard to read.
This commit cleans it up.

Epic: None

Release note: None

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130532)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @nvanbenschoten, and @pav-kv)

___
*[`pkg/raft/raft.go` line 895 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8H4H7b8Mc9rSWFgYjk:-O8Hn1y_3NvL9PDiCgh6:b-22m4mn) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/raft.go#L895)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

We're sending `MsgApp`s here though, not `MsgAppResp`s, so maybe the wording could be improved.

Since we're only sending `MsgApp` here, this code could only update `SentCommit`. The `MatchCommit` is not updated for a different reason - there is nothing that triggers a self-directed `MsgAppResp` except `MsgStorageAppend`s.
</blockquote></details>

I updated the wording (in both places that had the same wording). Hopefully the new wording is better.
___
*[`pkg/raft/testdata/async_storage_writes_append_aba_race.txt` line 392 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8H61Iu49W1qEPRF1I1:-O8HnaJLDO5dHAUMGHGx:b-qu6n9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/testdata/async_storage_writes_append_aba_race.txt#L392)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

It's still "heartbeat" , looking at messages below. But I understand that heartbeats will eventually be removed.

Maybe it would be prudent to not specify in this comment which message type exactly does it, just broadly say "hears of the term change and persists the new term".
</blockquote></details>

Good point. Done!
___
*[`pkg/raft/testdata/heartbeat_resp_recovers_from_probing.txt` line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130532#gh-1781719641:-O8HkNN38IaUxd6bwqDN:b-xnpc8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee9e30ede69a4217369a97c8e460f3df68c3d6b/pkg/raft/testdata/heartbeat_resp_recovers_from_probing.txt#L77)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Is there a benefit in mixing the two behaviours?

I'm still wondering why we can't make it an either/or behaviour. For example in `case MsgHeartbeatResp:` handler, could we:

```golang
if !r.fortificationTracker.FortificationEnabled() {
	pr.MsgAppProbesPaused = false
	r.maybeSendAppend(m.From)
}
```

Could place this as a helper next to `maybeUnpauseAndBcastAppend`, to make it clear that these two are mutually exclusive behaviours.

Same for the `RecentlyActive` field updates.
</blockquote></details>

Removing that from heartbeats work, but it causes even more test changes. I believe it's ~easier to understand the code this way. Especially that the PR that removes heartbeats if `FortificationEnabled()` is true is ready after this PR.
___
*[`pkg/raft/testdata/lagging_commit_no_store_liveness_support.txt` line 137 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8HPNP21ZQCrh1eZfTe:-O8Hkwrn7G_s7wO78MCo:b-8f0bwn) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/testdata/lagging_commit_no_store_liveness_support.txt#L137)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

I wouldn't expect heartbeats here though. Is this the next change - disabling `MsgHeartbeat` if store liveness is used?
</blockquote></details>

Correct. These will be removed soon.
We are now just in this not very clean state, but that all should get cleaned this week or next week (right after this PR is merged).
___
*[`pkg/raft/raft_flow_control_test.go` line 156 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8HWuNv2CcMlQQlw-Hw:-O8HvWaU0qpHdFWsm1P4:b-p35li3) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/raft_flow_control_test.go#L156)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

This many-fold nesting doesn't look good. Maybe this test could be cleaned up before/after being modified. Not necessary to do in this PR, just a thought. It's generally a good practice - always leave the place cleaner than it was before.
</blockquote></details>

I added a commit in the middle to clean this up a little
___
*[`pkg/raft/raft_test.go` line 2113 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8GyXMB2i6ufxqIkFF7:-O8HlOPx6w0UP9943WqT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6121b09099008d937b2408c9f4837d0dd1797beb/pkg/raft/raft_test.go#L2113)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

I wonder why this code needed to send a heartbeat `heartbeatTimeout` times. That doesn't seem right, perhaps it only worked because the `heartbeatTimeout == 1`, set in `newTestRaft` call.

The new code makes more sense.
</blockquote></details>

Done.
___
*[`pkg/raft/raft_test.go` line 2122 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8GxaFA9t7PO2xxfLDb:-O8Hp6CLA99Lf5gwUoiI:bawpaw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4be3c73e83a1ea66442e9c0585cdfa40a3fea90/pkg/raft/raft_test.go#L2122)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Pre-existing, but not a big fan of this verbose construct. Can it be done as:

```golang
r := newTestRaft(..., newTestMemoryStorage(...),
	withFortification(storeLivenessEnabled))
```
</blockquote></details>

I added a TODO for that as well. I am planning to address these type of TODOs in the release stability period.
___
*[`pkg/raft/raft_test.go` line 4084 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8H-LWK3bLpWJrLhXij:-O8HpaaH8B7yBMS065eR:b-7qmo62) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4be3c73e83a1ea66442e9c0585cdfa40a3fea90/pkg/raft/raft_test.go#L4084)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

`advanceMessagesAfterAppend` sounds like a thing that needs to be done after handling `Ready` and storing the log. Calling on tick is not necessarily right. But this harness is really confusing, would be great to clean it up at some point.

Looks like `p.readMessages()` calls into `p.advanceMessagesAfterAppend()`. Do we need to call both, or can just call the former?
</blockquote></details>

Good point. Done
___
*[`pkg/raft/raft_test.go` line 1225 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8HTvks1UCCY6Lmljql:-O8HqzN054I-42zOOYT6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/raft_test.go#L1225)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

I think the leader will be already at `Term: 4` at this time, because `becomeCandidate()` bumps the term, and then `becomeLeader()` appends a new entry at the new term, which bumps the `LogMark`'s term.

Unfortunately, we don't have good [safety checks](https://github.com/cockroachdb/cockroach/blob/290f68fb271d99f86faf7a7e7b1cb4087c7e8c65/pkg/raft/log.go#L364-L368) yet that would catch this here, but let's make sure we populate things correctly to save debugging later.

For now, can replace this with:

```golang
sm.raftLog.commitTo(sm.raftLog.unstable.mark())
```
</blockquote></details>

Done.
___
*[`pkg/raft/raft_test.go` line 1238 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8HVl1s1YPBt22LbTFh:-O8HoTwTEtMZ2xkhe6C1:bk1zm3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/raft_test.go#L1238)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Seeing that this repeated tick pattern is used a lot, consider adding a helper (maybe make it test-only, by placing in a `_test.go` file) like:

```golang
func (r *raft) tickTimes(times int) {
	// this loop
}
```

Or could leave a TODO for later.
</blockquote></details>

I added a TODO because there are many tests that could benefit from something like that, and I can do it in a refactoring PR later


<!-- Sent from Reviewable.io -->

pav-kv reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130532)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @iskettaneh, and @nvanbenschoten)

___
*[`pkg/raft/testdata/heartbeat_resp_recovers_from_probing.txt` line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130532#gh-1781719641:-O8IQpLv4B8G8DvXaThl:b-b58igm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee9e30ede69a4217369a97c8e460f3df68c3d6b/pkg/raft/testdata/heartbeat_resp_recovers_from_probing.txt#L77)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Removing that from heartbeats work, but it causes even more test changes. I believe it's ~easier to understand the code this way. Especially that the PR that removes heartbeats if `FortificationEnabled()` is true is ready after this PR.
</blockquote></details>

SGTM, don't want more test changes for now :)
___
*[`pkg/raft/raft_flow_control_test.go` line 156 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/130532#-O8HWuNv2CcMlQQlw-Hw:-O8IPy9sDeTwST7ChYl4:b4yrzfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/72889d6eec01a455d321fda15b09662ebcf50489/pkg/raft/raft_flow_control_test.go#L156)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

I added a commit in the middle to clean this up a little
</blockquote></details>

Appreciated, thanks!


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/130532)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @nvanbenschoten, and @pav-kv)

___
*[`pkg/raft/testdata/heartbeat_resp_recovers_from_probing.txt` line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/130532#gh-1781719641:-O8J0aOvAygqCYvYtJIZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ee9e30ede69a4217369a97c8e460f3df68c3d6b/pkg/raft/testdata/heartbeat_resp_recovers_from_probing.txt#L77)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

SGTM, don't want more test changes for now :)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

iskettaneh commented:
bors r+
